File: part1/python/closures_scope.py
# Section 2: Python closures & scope
def bad_funcs():
    fs = []
    for i in range(3):
        fs.append(lambda x: x + i)  # late binding
    return fs

def good_funcs():
    fs = []
    for i in range(3):
        fs.append(lambda x, i=i: x + i)  # capture value
    return fs

b = [f(10) for f in bad_funcs()]
g = [f(10) for f in good_funcs()]

print("Python bad (late-bound):", b)
print("Python good (captured):", g)

File: part1/javascript/cpp/closures_scope.cpp
// Section 2: C++ closures & capture
#include <iostream>
#include <vector>
#include <functional>
using namespace std;

vector<function<int(int)>> byRef() {
    vector<function<int(int)>> fs;
    int i;
    for (i = 0; i < 3; ++i) fs.push_back([&](int x){ return x + i; });
    return fs;
}
vector<function<int(int)>> byVal() {
    vector<function<int(int)>> fs;
    for (int i = 0; i < 3; ++i) fs.push_back([=](int x){ return x + i; });
    return fs;
}
int main() {
    auto r = byRef(); auto v = byVal();
    cout << "C++ [&] (late-bound): ";
    for (auto& f : r) cout << f(10) << " ";
    cout << "\nC++ [=] (captured):  ";
    for (auto& f : v) cout << f(10) << " ";
    cout << endl;
}
